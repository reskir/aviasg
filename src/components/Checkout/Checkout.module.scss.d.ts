export const element: string;
export const item: string;
export const summary: string;
