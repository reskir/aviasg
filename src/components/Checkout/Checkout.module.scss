.summary {
    display: grid;
    grid-row-gap: 32px;
    margin-bottom: 24px;
}

.element {
    display: grid;
    grid-row-gap: 12px;
}

.item {
    display: grid;
    grid-template-columns: 0.3fr 0.7fr;
}
