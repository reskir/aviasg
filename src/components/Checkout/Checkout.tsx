import { useEffect } from 'react';
import { useHistory } from 'react-router';
import { useAppSelector, useAppDispatch } from '../../store';
import {
    useGetProductsAndTaxInfoQuery,
    useBuyMutation,
    IUpdatedProduct,
} from '../../service';
import { setCheckoutIsSuccess, IContact } from '../../slices/main';

import styles from './Checkout.module.scss';

export const CheckoutSummary = ({
    products,
    contact,
    totalNetPrice,
    totalTaxes,
    totalPrice,
}: {
    products: IUpdatedProduct[];
    contact: IContact;
    totalNetPrice: number;
    totalTaxes: number;
    totalPrice: number;
}) => {
    return (
        <div className={styles.summary}>
            <div className={styles.element}>
                <h2>Products</h2>
                {products.map(({ id, title, price }) => {
                    return (
                        <div className={styles.item} key={id}>
                            <div>{title} </div>
                            <div>
                                <strong>{price.netTotal}</strong>{' '}
                                {price.currency}
                            </div>
                        </div>
                    );
                })}
            </div>
            <div className={styles.element}>
                <h2>Contacts</h2>
                <div className={styles.item}>
                    <div>Name:</div>
                    <div>
                        {contact.firstName} {contact.lastName}
                    </div>
                </div>
            </div>
            <div className={styles.element}>
                <h2>Price</h2>
                <div className={styles.item}>
                    <div>Product(s):</div>
                    <div>
                        <strong>{totalNetPrice}</strong> EUR
                    </div>
                </div>
                <div className={styles.item}>
                    <div>Taxes:</div>
                    <div>
                        <strong>{totalTaxes}</strong> EUR
                    </div>
                </div>
                <div className={styles.item}>
                    <div>Total:</div>
                    <div>
                        <strong>{totalPrice}</strong> EUR
                    </div>
                </div>
            </div>
        </div>
    );
};

export const Checkout = (): JSX.Element => {
    const history = useHistory();
    const dispatch = useAppDispatch();
    const { cart, contact } = useAppSelector((state) => state.main);
    const { data, isLoading } = useGetProductsAndTaxInfoQuery({});
    const products = data?.products || [];
    const taxInfo = data?.taxInfo;

    const [checkout, { isSuccess }] = useBuyMutation();

    useEffect(() => {
        if (isSuccess) {
            dispatch(setCheckoutIsSuccess(isSuccess));
            history.push('/success');
        }
    }, [isSuccess]);

    if (isLoading) {
        return <span>Loading...</span>;
    }

    const productsToBuy = products.filter(({ id }) => cart.includes(id));
    const totals = productsToBuy.reduce(
        ({ totalNetPrice, totalTaxes }, { price }) => ({
            totalNetPrice: totalNetPrice + price.netTotal,
            totalTaxes: totalTaxes + price.taxes,
        }),
        { totalNetPrice: 0, totalTaxes: 0 }
    );

    const totalNetPrice = Number(totals.totalNetPrice.toFixed(2));
    const totalTaxes = Number(totals.totalTaxes.toFixed(2));
    const totalPrice = Number((totalNetPrice + totalTaxes).toFixed(2));

    if (productsToBuy.length) {
        return (
            <>
                <CheckoutSummary
                    products={productsToBuy}
                    totalNetPrice={totalNetPrice}
                    totalTaxes={totalTaxes}
                    totalPrice={totalPrice}
                    contact={contact}
                />
                <button
                    onClick={() =>
                        checkout({
                            products: cart,
                            contact,
                            price: {
                                netTotal: totalNetPrice,
                                taxes: totalTaxes,
                                grossTotal: totalPrice,
                                currency: 'EUR',
                            },
                            taxInfo,
                        })
                    }
                >
                    Complete order
                </button>
            </>
        );
    }

    return null;
};

export default Checkout;
