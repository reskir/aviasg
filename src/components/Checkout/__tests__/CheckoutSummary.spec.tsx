import React from 'react';
import renderer from 'react-test-renderer';
import { CheckoutSummary } from '../Checkout';
import { products, contact } from './data';

test('renders checkout summary correctly', async () => {
    const component = renderer
        .create(
            <CheckoutSummary
                products={products}
                contact={contact}
                totalNetPrice={20.34}
                totalPrice={30.34}
                totalTaxes={10}
            />
        )
        .toJSON();
    expect(component).toMatchSnapshot();
});
