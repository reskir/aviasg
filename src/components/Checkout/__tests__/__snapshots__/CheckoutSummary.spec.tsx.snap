// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders checkout summary correctly 1`] = `
<div
  className="summary"
>
  <div
    className="element"
  >
    <h2>
      Products
    </h2>
    <div
      className="item"
    >
      <div>
        Product 1
         
      </div>
      <div>
        <strong>
          10.49
        </strong>
         
        EUR
      </div>
    </div>
    <div
      className="item"
    >
      <div>
        Product 2
         
      </div>
      <div>
        <strong>
          11.19
        </strong>
         
        EUR
      </div>
    </div>
    <div
      className="item"
    >
      <div>
        Product 3
         
      </div>
      <div>
        <strong>
          13.69
        </strong>
         
        EUR
      </div>
    </div>
    <div
      className="item"
    >
      <div>
        Product 4
         
      </div>
      <div>
        <strong>
          17.29
        </strong>
         
        EUR
      </div>
    </div>
  </div>
  <div
    className="element"
  >
    <h2>
      Contacts
    </h2>
    <div
      className="item"
    >
      <div>
        Name:
      </div>
      <div>
        Test
         
        Test
      </div>
    </div>
  </div>
  <div
    className="element"
  >
    <h2>
      Price
    </h2>
    <div
      className="item"
    >
      <div>
        Product(s):
      </div>
      <div>
        <strong>
          20.34
        </strong>
         EUR
      </div>
    </div>
    <div
      className="item"
    >
      <div>
        Taxes:
      </div>
      <div>
        <strong>
          10
        </strong>
         EUR
      </div>
    </div>
    <div
      className="item"
    >
      <div>
        Total:
      </div>
      <div>
        <strong>
          30.34
        </strong>
         EUR
      </div>
    </div>
  </div>
</div>
`;
